### PR TITLE
combine fcp_list into one punch script

### PR DESCRIPTION
**Pull request by SharpRazor**
before this PR, we need a loop to dedicate and configure each FCP in fcp list assigned for the target linux. And for each FCP, we need punch a script for it.

Now we use this commit to combine fcp list into one punch script, we only need one script to complete the configuration of attachment or detachment.
Signed-off-by: SharpRazor <bjcb@cn.ibm.com>

### CODE REVIEW

File: zvmsdk/volumeop.py [hunk 1]
```diff
@@ -173,23 +172,23 @@ def config_attach(self, fcp, assigner_id, target_wwpns, target_lun,
                 if exit_code == 1:
                     errmsg = ('attach scripts execution failed because the '
                               'device %s did not show up in the target '
-                              'virtual machine %s , please check its '
-                              'connections or settings.' % (fcp, assigner_id))
+                              'machine %s , please check its connections '
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:175]: Error msg is not accurate: device %s (fcp_list) did not show up 
should be the disk does not show up in the target virtual machine (not fcp_list)

File: zvmsdk/volumeop.py [hunk 2]
```diff
@@ -152,13 +152,12 @@ def _get_status_code_from_systemctl(self, assigner_id, command):
...
         return exit_code
 
-    def config_attach(self, fcp, assigner_id, target_wwpns, target_lun,
-                      multipath, os_version, mount_point, new,
-                      need_restart=True):
+    def config_attach(self, fcp_list, assigner_id, target_wwpns, target_lun,
+                      multipath, os_version, mount_point, need_restart=True):
         linuxdist = self._dist_manager.get_linux_dist(os_version)()
-        self.configure_volume_attach(fcp, assigner_id, target_wwpns,
+        self.configure_volume_attach(fcp_list, assigner_id, target_wwpns,
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:158]: suggest to log the parameters as info log.

File: zvmsdk/volumeop.py [hunk 3]
```diff
@@ -173,23 +172,23 @@ def config_attach(self, fcp, assigner_id, target_wwpns, target_lun,
...
                                                         userid=assigner_id,
                                                         msg=errmsg)
 
-    def config_detach(self, fcp, assigner_id, target_wwpns, target_lun,
+    def config_detach(self, fcp_list, assigner_id, target_wwpns, target_lun,
                       multipath, os_version, mount_point, connections,
                       need_restart=True):
         linuxdist = self._dist_manager.get_linux_dist(os_version)()
-        self.configure_volume_detach(fcp, assigner_id, target_wwpns,
+        self.configure_volume_detach(fcp_list, assigner_id, target_wwpns,
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:191]: log.info key parameters

File: zvmsdk/volumeop.py [hunk 4]
```diff
@@ -758,30 +757,51 @@ def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
...
         call smt layer to call linux command
         """
-        LOG.info('Start to attach device %s to %s' % (fcp, assigner_id))
+        LOG.info("Start to attach devices %s to "
+                 "%s" % (fcp_list, assigner_id))
+
+        if is_root_volume:
+            LOG.info("Attaching devices %s to %s is "
+                     "done." % (fcp_list, assigner_id))
+            # root volume will be dedicated by refresh_bootmap
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py]: this comment is wrong.  the FCP devices are defined in user entry, don't need to use smcli to dedicate.

File: zvmsdk/volumeop.py [hunk 5]
```diff
@@ -826,6 +846,12 @@ def attach(self, connection_info):
         mount_point = connection_info['mount_point']
         is_root_volume = connection_info.get('is_root_volume', False)
 
+        if len(fcp) == 0:
+            errmsg = ("The passed in fcps in connection_info %s "
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:850]: why would you add this check? I cann't think of a scenario for this to happen.

File: zvmsdk/volumeop.py [hunk 6]
```diff
@@ -826,6 +846,12 @@ def attach(self, connection_info):
...
+                      "is empty when enter attach." % connection_info)
+            raise exception.SDKVolumeOperationError(rs=7,
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:852]: Also rs=7 is for refreshbootmap error, not this.

File: zvmsdk/volumeop.py [hunk 7]
```diff
@@ -835,57 +861,56 @@ def attach(self, connection_info):
...
-                                       os_version, mount_point, path_count,
-                                       is_root_volume, need_restart)
-                    if new and is_root_volume is False:
-                        dedicated_fcp.append(fcp[i])
-                except exception.SDKBaseException:
-                    self._rollback_dedicated_fcp(dedicated_fcp, assigner_id,
-                        all_fcp_list=fcp)
-                    raise
+            # always restart to check the return value of attachment
+            need_restart = True
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:865]: no need to keep the need_restart parameter, to keep our code clean.

File: zvmsdk/volumeop.py [hunk 8]
```diff
@@ -758,30 +757,51 @@ def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
...
-            self._add_disk(fcp, assigner_id, target_wwpns, target_lun,
-                           multipath, os_version, mount_point, new,
-                           need_restart)
+        dedicated_fcps = []
+        try:
+            # dedicate the new FCP devices to the userid
+            for fcp in fcp_list:
+                if fcp_status[fcp]:
+                    # only dedicate the ones first attached
+                    self._dedicate_fcp(fcp, assigner_id)
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:788]: 1. log.info ( "start to dedicate FCP xxx to userid: xxx.") before _dedicate_fcp
2. log.info (FCP xxx dedicated to xxx) after _dedicate_fcp

File: zvmsdk/volumeop.py [hunk 9]
```diff
@@ -758,30 +757,51 @@ def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
         dedicate fcp to the user if it's needed, after that
         call smt layer to call linux command
         """
-        LOG.info('Start to attach device %s to %s' % (fcp, assigner_id))
+        LOG.info("Start to attach devices %s to "
+                 "%s" % (fcp_list, assigner_id))
+
+        if is_root_volume:
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py]: you cann't just return at the beginning.
for root_volume, should return after connections is increased which is done in add_fcp_for_assigner.

File: zvmsdk/volumeop.py [hunk 10]
```diff
@@ -758,30 +757,51 @@ def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
...
+            LOG.info("Attaching devices %s to %s is "
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py]: the log is confusing, should say sth like "is root volume, xxx."

File: zvmsdk/volumeop.py [hunk 11]
```diff
@@ -758,30 +757,51 @@ def _attach(self, fcp, assigner_id, target_wwpns, target_lun,
...
+                  "dedicating them to %s is: %s." % (assigner_id, fcp_status))
 
-            self._add_disk(fcp, assigner_id, target_wwpns, target_lun,
-                           multipath, os_version, mount_point, new,
-                           need_restart)
+        dedicated_fcps = []
+        try:
+            # dedicate the new FCP devices to the userid
+            for fcp in fcp_list:
+                if fcp_status[fcp]:
```
Comment (dongyanyang) [inline zvmsdk/volumeop.py:786]: add a else for this, to log.info("This is not the first volume for FCP xxx, skip dedicating FCP device.")

**Review by dongyanyang — CHANGES_REQUESTED** [review]

File: zvmsdk/volumeops/templates/rhel7_attach_volume.j2 [hunk 12]
```diff
@@ -67,44 +43,97 @@ all_hosts=(`ls /sys/class/scsi_host/`)
...
+    Timeout=60
+    while [ $FoundDiskPath -eq 0 ]
+    do
+        # if timeout less or equal 0 seconds, means no time left
+        if [ $Timeout -le 0 ];then
+            echo "waiting for devices ready timed out after 60 seconds."
             break
         fi
+        # loop all the WWPNs to found the alive device
+        for j in ${ActiveWWPNs[@]}
```
Comment (dongyanyang) [inline zvmsdk/volumeops/templates/rhel7_attach_volume.j2:110]: should move the device check out of the current loop, after all FCPs are done unit_add.

File: zvmsdk/volumeops/templates/rhel7_detach_volume.j2 [hunk 13]
```diff
@@ -1,87 +1,97 @@
...
-# get the real WWPNs in the file system
-declare -a ActiveWWPNs
-ActiveWWPNs=(`ls /sys/bus/ccw/drivers/zfcp/0.0.$fcp/ | grep "0x"`)
+declare -a WWIDs
+index=0
+for device in ${SourceDevices[@]}
+do
+    wwid=$(/lib/udev/scsi_id --page 0x83 --whitelisted /dev/disk/by-path/$device)
+    echo "scsi_id command get WWID:$wwid for device: $device under /dev/disk/by-path"
+    WWIDs[index]=$wwid
```
Comment (dongyanyang) [inline zvmsdk/volumeops/templates/rhel7_detach_volume.j2:22]: di'fferent path has same wwid. no need for array

**Review by dongyanyang — CHANGES_REQUESTED** [review]

File: zvmsdk/volumeop.py (hunk 1 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:175]: done.

File: zvmsdk/volumeop.py (hunk 2 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:158]: done, but add to config_attach.

File: zvmsdk/volumeop.py (hunk 3 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:191]: done.

File: zvmsdk/volumeop.py (hunk 4 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py]: changed, done.

File: zvmsdk/volumeop.py (hunk 5 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:850]: deleted, done.

File: zvmsdk/volumeop.py (hunk 7 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:865]: deleted, done.

File: zvmsdk/volumeop.py (hunk 8 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:788]: done.

File: zvmsdk/volumeop.py (hunk 9 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py]: oh, yes, done.

File: zvmsdk/volumeop.py (hunk 11 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py:786]: done.

File: zvmsdk/volumeop.py (hunk 10 above)
Comment (SharpRazor) [reply zvmsdk/volumeop.py]: updated,  done.

File: zvmsdk/volumeop.py [hunk 14]
```diff
@@ -923,23 +950,28 @@ def detach(self, connection_info):
         else:
             multipath = False
 
+        if len(fcp) == 0:
```
[the remark this answers is not in the document]
Comment (SharpRazor) [reply zvmsdk/volumeop.py:953]: deleted, done.

File: zvmsdk/volumeop.py [hunk 15]
```diff
@@ -923,23 +950,28 @@ def detach(self, connection_info):
...
-            for i in range(path_count):
-                # To avoid restart zvmguestconfigure service too often,
-                # only restart it when the last fcp is detached
-                if i == path_count - 1:
-                    need_restart = True
-                self._detach(fcp[i].lower(), assigner_id, target_wwpns,
-                             target_lun, multipath, os_version, mount_point,
-                             is_root_volume, need_restart)
+            # always restart to check the return value of detachment
+            need_restart = True
```
[the remark this answers is not in the document]
Comment (SharpRazor) [reply zvmsdk/volumeop.py:968]: done.

File: zvmsdk/volumeop.py [hunk 16]
```diff
@@ -835,57 +861,56 @@ def attach(self, connection_info):
...
                                       target_lun, multipath, os_version,
                                       mount_point, connections, need_restart)
 
-    def _detach(self, fcp, assigner_id, target_wwpns, target_lun,
+    def _detach(self, fcp_list, assigner_id, target_wwpns, target_lun,
                 multipath, os_version, mount_point, is_root_volume,
                 need_restart):
         """Detach a volume from a guest"""
-        LOG.info('Start to detach device %s from %s' % (fcp, assigner_id))
+        LOG.info("Start to detach device %s from "
```
[the remark this answers is not in the document]
Comment (SharpRazor) [reply zvmsdk/volumeop.py:888]: done, please help review the new message content again, thanks.

File: zvmsdk/volumeop.py [hunk 17]
```diff
@@ -835,57 +861,56 @@ def attach(self, connection_info):
...
+        # the values are the connections colume value in database
+        fcp_connections = {}
         try:
-            connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
+            for fcp in fcp_list:
+                connections = self.fcp_mgr.decrease_fcp_usage(fcp, assigner_id)
+                fcp_connections[fcp] = connections
         except exception.SDKObjectNotExistError:
             connections = 0
             LOG.warning("The connections of FCP device %s is 0.", fcp)
```
[the remark this answers is not in the document]
Comment (SharpRazor) [reply zvmsdk/volumeop.py:899]: this is not added by my code, I dont know the reason, so I kept it.

File: zvmsdk/volumeop.py [hunk 18]
```diff
@@ -835,57 +861,56 @@ def attach(self, connection_info):
...
-                              multipath, os_version, mount_point, connections,
-                              need_restart)
-            if not connections:
-                self._undedicate_fcp(fcp, assigner_id)
+            self._remove_disks(fcp_list, assigner_id, target_wwpns, target_lun,
+                               multipath, os_version, mount_point, connections,
+                               need_restart)
+            for fcp in fcp_list:
+                if not fcp_connections[fcp]:
+                    self._undedicate_fcp(fcp, assigner_id)
```
[the remark this answers is not in the document]
Comment (SharpRazor) [reply zvmsdk/volumeop.py:912]: done.

File: zvmsdk/volumeops/templates/rhel7_attach_volume.j2 (hunk 12 above)
Comment (SharpRazor) [reply zvmsdk/volumeops/templates/rhel7_attach_volume.j2:110]: right, done.

File: zvmsdk/volumeops/templates/rhel7_detach_volume.j2 (hunk 13 above)
Comment (SharpRazor) [reply zvmsdk/volumeops/templates/rhel7_detach_volume.j2:22]: right, done.

**Comment by SharpRazor** [comment]
the log in zvmsdk.log for attaching 3 volumes:
```
[2021-04-21 08:00:45] [INFO] Reserve fcp device 1a13
[2021-04-21 08:00:45] [INFO] Reserve fcp device 1b14
[2021-04-21 08:00:46] [INFO] Got available fcp_list [('1a13', 'CBM5400C', 0, 1, 0, ''), ('1b14', 'CBM5400C', 0, 1, 1, '')] in Reserve mode from CBM5400C.
[2021-04-21 08:00:46] [INFO] Found allocated fcps [('1a13', 'CBM5400C', 0, 1, 0, ''), ('1b14', 'CBM5400C', 0, 1, 1, '')] for CBM5400C, will reuse them.
[2021-04-21 08:00:46] [INFO] Reserve fcp device 1a13
[2021-04-21 08:00:46] [INFO] Reserve fcp device 1b14
[2021-04-21 08:00:46] [INFO] Got available fcp_list [('1a13', 'CBM5400C', 0, 1, 0, ''), ('1b14', 'CBM5400C', 0, 1, 1, '')] in Reserve mode from CBM5400C.
[2021-04-21 08:00:46] [INFO] Found allocated fcps [('1a13', 'CBM5400C', 0, 1, 0, ''), ('1b14', 'CBM5400C', 0, 1, 1, '')] for CBM5400C, will reuse them.
[2021-04-21 08:00:46] [INFO] Reserve fcp device 1a13
[2021-04-21 08:00:46] [INFO] Reserve fcp device 1b14
[2021-04-21 08:00:47] [INFO] get_volume_connector returns {'zvm_fcp': ['1a13', '1b14'], 'wwpns': ['c05076de33000377', 'c05076de33000373'], 'host': 'BOEM5401'} for CBM5400C
[2021-04-21 08:00:48] [INFO] get_volume_connector returns {'zvm_fcp': ['1a13', '1b14'], 'wwpns': ['c05076de33000377', 'c05076de33000373'], 'host': 'BOEM5401'} for CBM5400C
[2021-04-21 08:00:48] [INFO] get_volume_connector returns {'zvm_fcp': ['1a13', '1b14'], 'wwpns': ['c05076de33000377', 'c05076de33000373'], 'host': 'BOEM5401'} for CBM5400C
[2021-04-21 08:00:48] [INFO] get_volume_connector returns {'zvm_fcp': ['1a13', '1b14'], 'wwpns': ['c05076de33000377', 'c05076de33000373'], 'host': 'BOEM5401'} for CBM5400C
[2021-04-21 08:00:51] [INFO] Start to attach volume to FCP devices ['1a13', '1b14'] on machine CBM5400C.
[2021-04-21 08:00:53] [INFO] Start to dedicate FCP 1a13 to CBM5400C.
[2021-04-21 08:00:54] [INFO] Start to attach volume to FCP devices ['1a13', '1b14'] on machine CBM5400C.
[2021-04-21 08:00:55] [INFO] Start to attach volume to FCP devices ['1a13', '1b14'] on machine CBM5400C.
[2021-04-21 08:00:55] [INFO] FCP 1a13 dedicated to CBM5400C isdone.
[2021-04-21 08:00:55] [INFO] Start to dedicate FCP 1b14 to CBM5400C.
[2021-04-21 08:00:56] [INFO] This is not the first volume for FCP 1a13, skip dedicating FCP device.
[2021-04-21 08:00:56] [INFO] This is not the first volume for FCP 1b14, skip dedicating FCP device.
[2021-04-21 08:00:56] [INFO] Begin to configure volume (WWPN:['0x5005076306031388', '0x5005076306035388', '0x5005076306039388', '0x500507630603D388', '0x5005076306081388', '0x5005076306085388', '0x5005076306089388', '0x500507630608D388', '0x5005076306101388', '0x5005076306105388', '0x5005076306109388', '0x500507630610D388', '0x5005076306131388', '0x5005076306135388', '0x5005076306139388', '0x500507630613D388', '0x5005076306181388', '0x5005076306185388', '0x5005076306189388', '0x500507630618D388', '0x50050763061B1388', '0x50050763061B5388', '0x50050763061B9388', '0x50050763061BD388'], LUN:0x4001403800000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'].
[2021-04-21 08:00:57] [INFO] This is not the first volume for FCP 1a13, skip dedicating FCP device.
[2021-04-21 08:00:57] [INFO] This is not the first volume for FCP 1b14, skip dedicating FCP device.
[2021-04-21 08:00:57] [INFO] Begin to configure volume (WWPN:['0x500507680B24BAC6', '0x500507680B23BAC6', '0x500507680B21BAC6', '0x500507680B22BAC6', '0x500507680B24BAC7', '0x500507680B22BAC7', '0x500507680B21BAC7', '0x500507680B23BAC7'], LUN:0x0001000000000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'].
[2021-04-21 08:00:58] [INFO] FCP 1b14 dedicated to CBM5400C isdone.
[2021-04-21 08:00:58] [INFO] Begin to configure volume (WWPN:['0x500507680B24BAC6', '0x500507680B23BAC6', '0x500507680B21BAC6', '0x500507680B22BAC6', '0x500507680B24BAC7', '0x500507680B22BAC7', '0x500507680B21BAC7', '0x500507680B23BAC7'], LUN:0x0000000000000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'].
[2021-04-21 08:01:04] [INFO] Start to attach volume to FCP devices ['1a13', '1b14'] on machine CBM5400C.
[2021-04-21 08:01:06] [INFO] This is not the first volume for FCP 1a13, skip dedicating FCP device.
[2021-04-21 08:01:06] [INFO] This is not the first volume for FCP 1b14, skip dedicating FCP device.
[2021-04-21 08:01:06] [INFO] Begin to configure volume (WWPN:['0x5005076306031388', '0x5005076306035388', '0x5005076306039388', '0x500507630603D388', '0x5005076306081388', '0x5005076306085388', '0x5005076306089388', '0x500507630608D388', '0x5005076306101388', '0x5005076306105388', '0x5005076306109388', '0x500507630610D388', '0x5005076306131388', '0x5005076306135388', '0x5005076306139388', '0x500507630613D388', '0x5005076306181388', '0x5005076306185388', '0x5005076306189388', '0x500507630618D388', '0x50050763061B1388', '0x50050763061B5388', '0x50050763061B9388', '0x50050763061BD388'], LUN:0x4015400800000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'].
[2021-04-21 08:01:15] [INFO] Configuration of volume (WWPN:['0x5005076306031388', '0x5005076306035388', '0x5005076306039388', '0x500507630603D388', '0x5005076306081388', '0x5005076306085388', '0x5005076306089388', '0x500507630608D388', '0x5005076306101388', '0x5005076306105388', '0x5005076306109388', '0x500507630610D388', '0x5005076306131388', '0x5005076306135388', '0x5005076306139388', '0x500507630613D388', '0x5005076306181388', '0x5005076306185388', '0x5005076306189388', '0x500507630618D388', '0x50050763061B1388', '0x50050763061B5388', '0x50050763061B9388', '0x50050763061BD388'], LUN:0x4001403800000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'] is done.
[2021-04-21 08:01:15] [INFO] Attaching volume to FCP devices ['1a13', '1b14'] on machine CBM5400C is done.
[2021-04-21 08:01:16] [INFO] Configuration of volume (WWPN:['0x500507680B24BAC6', '0x500507680B23BAC6', '0x500507680B21BAC6', '0x500507680B22BAC6', '0x500507680B24BAC7', '0x500507680B22BAC7', '0x500507680B21BAC7', '0x500507680B23BAC7'], LUN:0x0001000000000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'] is done.
[2021-04-21 08:01:16] [INFO] Attaching volume to FCP devices ['1a13', '1b14'] on machine CBM5400C is done.
[2021-04-21 08:01:16] [INFO] Configuration of volume (WWPN:['0x5005076306031388', '0x5005076306035388', '0x5005076306039388', '0x500507630603D388', '0x5005076306081388', '0x5005076306085388', '0x5005076306089388', '0x500507630608D388', '0x5005076306101388', '0x5005076306105388', '0x5005076306109388', '0x500507630610D388', '0x5005076306131388', '0x5005076306135388', '0x5005076306139388', '0x500507630613D388', '0x5005076306181388', '0x5005076306185388', '0x5005076306189388', '0x500507630618D388', '0x50050763061B1388', '0x50050763061B5388', '0x50050763061B9388', '0x50050763061BD388'], LUN:0x4015400800000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'] is done.
[2021-04-21 08:01:16] [INFO] Attaching volume to FCP devices ['1a13', '1b14'] on machine CBM5400C is done.
[2021-04-21 08:01:16] [INFO] Configuration of volume (WWPN:['0x500507680B24BAC6', '0x500507680B23BAC6', '0x500507680B21BAC6', '0x500507680B22BAC6', '0x500507680B24BAC7', '0x500507680B22BAC7', '0x500507680B21BAC7', '0x500507680B23BAC7'], LUN:0x0000000000000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'] is done.
[2021-04-21 08:01:16] [INFO] Attaching volume to FCP devices ['1a13', '1b14'] on machine CBM5400C is done.

```

**Comment by SharpRazor** [comment]
the log in zvmsdk.log for detaching 1 volumes after applying this PR:
```
[2021-04-21 08:09:20] [INFO] Start to detach volume on machine CBM5400C from FCP devices ['1a13', '1b14']
[2021-04-21 08:09:20] [INFO] Begin to deconfigure volume (WWPN:['0x500507680B24BAC6', '0x500507680B23BAC6', '0x500507680B21BAC6', '0x500507680B22BAC6', '0x500507680B24BAC7', '0x500507680B22BAC7', '0x500507680B21BAC7', '0x500507680B23BAC7'], LUN:0x0000000000000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'].
[2021-04-21 08:09:21] [INFO] Deconfiguration of volume (WWPN:['0x500507680B24BAC6', '0x500507680B23BAC6', '0x500507680B21BAC6', '0x500507680B22BAC6', '0x500507680B24BAC7', '0x500507680B22BAC7', '0x500507680B21BAC7', '0x500507680B23BAC7'], LUN:0x0000000000000000) on the target machine CBM5400C with FCP devices ['1a13', '1b14'] is done.
[2021-04-21 08:09:21] [INFO] Found still have volumes on FCP 1a13, skip undedicating FCP device.
[2021-04-21 08:09:21] [INFO] Found still have volumes on FCP 1b14, skip undedicating FCP device.
[2021-04-21 08:09:21] [INFO] Detaching volume on machine CBM5400C from FCP devices ['1a13', '1b14'] is done.
[2021-04-21 08:09:23] [INFO] Got available fcp_list [('1a13', 'CBM5400C', 3, 1, 0, ''), ('1b14', 'CBM5400C', 3, 1, 1, '')] in Unreserve mode from CBM5400C.
[2021-04-21 08:09:25] [INFO] get_volume_connector returns {'zvm_fcp': ['1a13', '1b14'], 'wwpns': ['c05076de33000377', 'c05076de33000373'], 'host': 'BOEM5401'} for CBM5400C
```

**Review by dongyanyang — APPROVED** [review]

**Comment by dongyanyang** [comment]
@jichenjc @bjhuangr please help to review and merge, thanks!

**Pull request merged**